contracts/extensions/version-control/ConditionalVersionControl.sol

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { ERC1967Upgrade } from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Upgrade.sol";
+import { IConditionalVersionControl } from "../../interfaces/version-control/IConditionalVersionControl.sol";
+import { ErrorHandler } from "../../libraries/ErrorHandler.sol";
+import { AddressArrayUtils } from "../../libraries/AddressArrayUtils.sol";
+import { ErrOnlySelfCall } from "../../utils/CommonErrors.sol";
+
+/**
+ * @title ConditionalVersionControl
+ * @dev A contract that allows conditional version control of contract implementations.
+ */
+abstract contract ConditionalVersionControl is IConditionalVersionControl, ERC1967Upgrade {
+  using ErrorHandler for bool;
+  using AddressArrayUtils for address[];
+
+  address public immutable proxyStorage;
+  address public immutable newVersion;
+  address public immutable currentVersion;
+
+  /**
+   * @dev Modifier that only allows self calls.
+   */
+  modifier onlySelfCall() {
+    _requireSelfCall();
+    _;
+  }
+
+  /**
+   * @dev Modifier that executes the function when conditions are met.
+   */
+  modifier whenConditionsAreMet() virtual;
```

Comment:
`whenConditionsAreMet --> atSwitchPoint`

Ref: https://en.wikipedia.org/wiki/Railroad_switch

Reply:
and internal checking methods could be `shouldDispatchNewImpl` or `shouldDispatchOldImpl`

Reply:
`whenConditionsAreMet` is quite general IMO, which "conditions" to do what "actions"?